### PR TITLE
Add tasks for poisoning and checking for poison in packages.

### PR DESCRIPTION
**Pull request by crummel**
Still having some build errors locally, so I want to see what this does on all legs.

### CODE REVIEW

**Comment by crummel** [comment]
@dagood and @dseefeld - this isn't completely ready yet, I'm still adding usage of the tasks to the build, but the tasks should be ready for a first look if you have time.

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CatalogFileEntry.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        const string ElementName = "File";
+
+        internal string Path { get; set; }
+        internal byte[] OriginalHash { get; set; }
+        internal byte[] PoisonedHash { get; set; }
+
+        public XElement ToXml() => new XElement(ElementName,
+            new XAttribute(nameof(Path), Path),
+            new XAttribute(nameof(OriginalHash), OriginalHash.ToHexString()),
+            PoisonedHash == null ? null : new XAttribute(nameof(PoisonedHash), PoisonedHash.ToHexString())
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CatalogFileEntry.cs]: Use `ToXAttributeIfNotNull` on `PoisonedHash?.ToHexString()`?
https://github.com/dotnet/source-build/blob/79b70dc0f353d2e652d83ee91936a7d74dce3cae/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/XmlParsingHelpers.cs#L21-L22

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CatalogFileEntry.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CatalogFileEntry.cs]: Missing copyright headers.

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        /// The marker file name to check for in poisoned nupkg files.
+        /// </summary>
+        public string MarkerFileName { get; set; }
+
+        /// <summary>
+        /// If true, fails the build if any poisoned files are found.
+        /// </summary>
+        public bool FailOnPoisonFound { get; set; }
+
+        static readonly string[] ZipFileExtensions =
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs]: Missing visibility specifier.

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            ".woff2",
+            ".xaml",
+            ".xml",
+        };
+
+        const string PoisonMarker = "POISONED";
+
+        public override bool Execute()
+        {
+            var poisons = GetPoisonedFiles(FilesToCheck.Select(f => f.ItemSpec), HashCatalogFilePath, MarkerFileName);
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs]: A bunch of non-obvious `var`s in this file.

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+            {
+                Log.LogWarning($"{poisons.Count()} marked files leaked to output.  See complete report '{PoisonReportOutputFilePath}' for details.");
+            }
+            else
+            {
+                Log.LogError($"No leaked files found in output.  Either something is broken or it is the future and we have fixed all leaks - please verify and remove this error if so (and default FailOnPoisonFound to true).");
+                return false;
+            }
+
+            return true;
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs]: Use `return !Log.HasLoggedError` (conventional).

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+                    fEntry.PoisonedHash = f.Attributes["PoisonedHash"]?.Value?.ToBytes();
+                    fEntry.Path = f.Attributes["Path"].Value;
+                }
+            }
+            return packages;
+        }
+    }
+
+    [Flags]
+    internal enum PoisonType
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs]: Should be in its own file, or be a private nested class.

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/PoisonedFileEntry.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        internal string Path { get; set; }
+        internal PoisonType Type { get; set; }
+        internal List<PoisonMatch> Matches { get; }
+
+        internal PoisonedFileEntry()
+        {
+            this.Matches = new List<PoisonMatch>();
+        }
+
+        public virtual XElement ToXml() => this.ToXml(ElementName);
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/PoisonedFileEntry.cs]: Why `virtual`?

File: tools-local/init-build.proj [hunk 8]
```diff
@@ -93,4 +94,13 @@
     <WriteLinesToFile File="$(TargetInfoProps)" Lines="$(TargetInfoPropsContent)" Overwrite="True" />
   </Target>
 
+  <Target Name="PoisonPrebuiltPackages" Condition="'$(EnablePoison)' == 'true' and '$(OfflineBuild)' == 'true'">
+    <ItemGroup>
+      <PrebuiltPackages Include="$(PrebuiltPackagesPath)/**/*.nupkg" />
+      <PrebuiltSourceBuiltPackages Include="$(PrebuiltSourceBuiltPackagesPath)/**/*.nupkg" />
```
Comment (dagood) [inline tools-local/init-build.proj]: The `/` after the variable isn't needed, dir properties end in a dir separator conventionally.

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+            }
+
+            // first check for a matching poisoned or non-poisoned hash match:
+            // - non-poisoned is a potential error where the package was redownloaded.
+            // - poisoned is a use of a local package we were not expecting.
+            foreach (var matchingCatalogedPackage in catalogedPackages.Where(c => c.OriginalHash.SequenceEqual(poisonEntry.Hash) || (c.PoisonedHash?.SequenceEqual(poisonEntry.Hash) ?? false)))
+            {
+                poisonEntry.Type |= PoisonType.Hash;
+                var match = new PoisonMatch();
+                match.Package = matchingCatalogedPackage.Path;
```
Comment (dagood) [inline tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CheckForPoison.cs]: Use object initializer? (More throughout.)

**Comment by crummel** [comment]
@dseefeld I think I've addressed most of Davis' comments if you want to take another look.

File: build-source-tarball.sh [hunk 10]
```diff
@@ -16,6 +16,7 @@ TARBALL_ROOT=$1
 shift
 
 SKIP_BUILD=0
+INCLUDE_LEAK_DETECTION=1
```
Comment (crummel) [inline build-source-tarball.sh]: Will change this to default to 0 before checkin.

File: support/tarball/build.sh [hunk 11]
```diff
@@ -11,7 +11,7 @@ export DOTNET_CLI_TELEMETRY_OPTOUT=1
 export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
 export NUGET_PACKAGES="$SCRIPT_ROOT/packages/"
 
-MSBUILD_ARGUMENTS=("/p:OfflineBuild=true" "/flp:v=detailed" "/clp:v=detailed")
+MSBUILD_ARGUMENTS=("/p:OfflineBuild=true" "/p:EnablePoison=true" "/flp:v=detailed" "/clp:v=detailed")
```
Comment (crummel) [inline support/tarball/build.sh]: Will change this to default to poison disabled before checkin.

File: tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CatalogFileEntry.cs (hunk 1 above)
Comment (crummel) [reply tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/CatalogFileEntry.cs]: Moved this out of the assembly so this is not available anymore.

File: build-source-tarball.sh [hunk 12]
```diff
@@ -158,6 +162,13 @@ SOURCE_BUILT_SDK_TOOLS_DIR="$TARBALL_ROOT/Tools/source-built/$ROSLYN_TOOLS_PACKA
 cp "$REPO_TOOLSET_PACKAGE_DIR/tools/"*.props "$SOURCE_BUILT_SDK_TOOLS_DIR"
 cp "$REPO_TOOLSET_PACKAGE_DIR/tools/"*.targets "$SOURCE_BUILT_SDK_TOOLS_DIR"
 
+if [ $INCLUDE_LEAK_DETECTION -eq 1 ]; then
+  find $TARBALL_ROOT/prebuilt
+  echo 'Building leak detection MSBuild tasks...'
+  DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1 ./Tools/dotnetcli/dotnet restore $SCRIPT_ROOT/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection.csproj --source $TARBALL_ROOT/prebuilt/source-built --source $TARBALL_ROOT/prebuilt/nuget-packages
+  DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1 ./Tools/dotnetcli/dotnet publish -o $FULL_TARBALL_ROOT/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection $SCRIPT_ROOT/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection.csproj
```
[the remark this answers is not in the document]
Comment (crummel) [reply build-source-tarball.sh]: Moved to cover the whole file.

**Comment by crummel** [comment]
@dotnet-bot test OSX10.12 Release please (java.nio.channels.ClosedChannelException)

**Pull request merged**